File: easylib/easylib.cpp
#include "easylib.h"
#include <windows.h>
#include <stdio.h>
#include "FileMapStream.h"
#include "MultiSymbol.h"
#include "ShapeLayer.h"

namespace easymap
{

BOOL APIENTRY DllMain(HANDLE hModule, DWORD ul_reason_for_call, LPVOID lpReserved)
{
    return TRUE;
}


bool DEF_EXPORT WINAPI CreateObj(const char* classname, IObj** ppObj)
{
    if (_invalid(ppObj)) return false;
    assert(!*ppObj);

    IObjPtr p;
    _FactoryManager::CreateInstance(classname, p);
    if (!p.Assigned()) return false;
    *ppObj = (IObj*)p._p();
    (*ppObj)->_AddRef();
    return true;
}

dword DEF_EXPORT WINAPI Instantiate(IStreamX* pStream, IPersist** ppPersist)
{
    if (_invalid(pStream) || _invalid(ppPersist)) return 0;
    assert(!*ppPersist);

    CStreamPtr ps = (CStream*)pStream;
    CPersistPtr pp;
    dword r = CPersist::Instantiate(ps, pp);
    if (pp.Assigned())
    {
        *ppPersist = (IPersist*)pp._p();
        (*ppPersist)->_AddRef();
    }
    return r;
}

bool DEF_EXPORT WINAPI CreateFileMappingStream(const char* filename,
    const bool readonly, IStreamX** ppStream)
{
    if (_invalid(filename) || _invalid(ppStream)) return false;
    assert(!*ppStream);

    CFileMapStreamPtr pFileMapStream = new CFileMapStream(filename, readonly);
    if (INVALID_HANDLE_VALUE == pFileMapStream->GetMapFileHandle()) return false;
    *ppStream = (IStreamX*)pFileMapStream._p();
    (*ppStream)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI CreateGeometry(easymap::GeometryType geometrytype,
    IGeometry** ppGeometry)
{
    if (_invalid(ppGeometry)) return false;
    assert(!*ppGeometry);

    IGeometryPtr pGeometry;
    switch (geometrytype)
    {
        case GEOMETRYTYPE_POINT: pGeometry = (IGeometry*)(new CPoint);
            break;
        case GEOMETRYTYPE_MULTIPOINT: pGeometry = (IGeometry*)(new CMultiPoint);
            break;
        case GEOMETRYTYPE_PATH: pGeometry = (IGeometry*)(new CPath);
            break;
        case GEOMETRYTYPE_RING: pGeometry = (IGeometry*)(new CRing);
            break;
        case GEOMETRYTYPE_POLYLINE: pGeometry = (IGeometry*)(new CPolyline);
            break;
        case GEOMETRYTYPE_POLYGON: pGeometry = (IGeometry*)(new CPolygon);
            break;
        case GEOMETRYTYPE_ENVELOPE: pGeometry = (IGeometry*)(new CEnvelope);
            break;
        case GEOMETRYTYPE_CIRCLE: pGeometry = (IGeometry*)(new CCircle);
            break;
        case GEOMETRYTYPE_ELLIPSE: pGeometry = (IGeometry*)(new CEllipse);
            break;
        default: return false;
    }

    *ppGeometry = pGeometry._p();
    (*ppGeometry)->_AddRef();
    return true;
}

dword DEF_EXPORT WINAPI Geometry2Stream(const IGeometry* pGeometry, IStreamX* pStream)
{
    if (_invalid(pGeometry) || _invalid(pStream)) return 0;
    IGeometryPtr pg = (IGeometry*)pGeometry;
    CStreamPtr ps = (CStream*)pStream;
    return Geometry2Stream(pg, ps);
}

dword DEF_EXPORT WINAPI Stream2Geometry(IStreamX* pStream, IGeometry** ppGeometry)
{
    if (_invalid(pStream) || _invalid(ppGeometry)) return 0;
    assert(!*ppGeometry);

    CStreamPtr ps = (CStream*)pStream;
    IGeometryPtr pg;
    dword r = Stream2Geometry(ps, pg);
    if (pg.Assigned())
    {
        *ppGeometry = pg._p();
        (*ppGeometry)->_AddRef();
    }
    return r;
}

bool DEF_EXPORT WINAPI CreateSimplePointSymbol(const double diameter, const COLORREF color,
    ISimplePointSymbol** ppSimplePointSymbol)
{
    if (_invalid(ppSimplePointSymbol)) return false;
    *ppSimplePointSymbol = new CSimplePointSymbol();
    (*ppSimplePointSymbol)->_AddRef();
    (*ppSimplePointSymbol)->SetDiameter(diameter);
    (*ppSimplePointSymbol)->SetColor(color);
    return true;
}

bool DEF_EXPORT WINAPI CreateEnvelopePointSymbol(const double width, const double height,
    const COLORREF color, IEnvelopePointSymbol** ppEnvelopePointSymbol)
{
    if (_invalid(ppEnvelopePointSymbol)) return false;
    assert(!*ppEnvelopePointSymbol);

    *ppEnvelopePointSymbol = new CEnvelopePointSymbol();
    (*ppEnvelopePointSymbol)->_AddRef();
    (*ppEnvelopePointSymbol)->SetWidth(width);
    (*ppEnvelopePointSymbol)->SetHeight(height);
    (*ppEnvelopePointSymbol)->SetColor(color);
    return true;
}

bool DEF_EXPORT WINAPI CreatePolyPointSymbol(const bool solid, const double linewidth,
    const COLORREF color, IPolyPointSymbol** ppPolyPointSymbol)
{
    if (_invalid(ppPolyPointSymbol)) return false;
    assert(!*ppPolyPointSymbol);

    *ppPolyPointSymbol = new CPolyPointSymbol;
    (*ppPolyPointSymbol)->_AddRef();
    (*ppPolyPointSymbol)->SetSolid(solid);
    (*ppPolyPointSymbol)->SetLineWidth(linewidth);
    (*ppPolyPointSymbol)->SetColor(color);
    return true;
}

bool DEF_EXPORT WINAPI CreateSimpleLineSymbol(const double width, const COLORREF color,
    ISimpleLineSymbol** ppSimpleLineSymbol)
{
    if (_invalid(ppSimpleLineSymbol)) return false;
    assert(!*ppSimpleLineSymbol);

    *ppSimpleLineSymbol = new CSimpleLineSymbol;
    (*ppSimpleLineSymbol)->_AddRef();
    (*ppSimpleLineSymbol)->SetWidth(width);
    (*ppSimpleLineSymbol)->SetColor(color);
    return true;
}

bool DEF_EXPORT WINAPI CreatePointLineSymbol(IPointLineSymbol** ppPointLineSymbol)
{
    if (_invalid(ppPointLineSymbol)) return false;
    assert(!*ppPointLineSymbol);

    *ppPointLineSymbol = new CPointLineSymbol;
    (*ppPointLineSymbol)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI CreateSimpleFillSymbol(const COLORREF color, const double borderwidth,
    const COLORREF bordercolor, ISimpleFillSymbol** ppSimpleFillSymbol)
{
    if (_invalid(ppSimpleFillSymbol)) return false;
    assert(!*ppSimpleFillSymbol);

    *ppSimpleFillSymbol = new CSimpleFillSymbol;
    (*ppSimpleFillSymbol)->_AddRef();
    (*ppSimpleFillSymbol)->SetColor(color);
    (*ppSimpleFillSymbol)->SetBorderWidth(borderwidth);
    (*ppSimpleFillSymbol)->SetBorderColor(bordercolor);
    return true;
}

bool DEF_EXPORT WINAPI CreatePointFillSymbol(IPointFillSymbol** ppPointFillSymbol,
    const COLORREF color, const double spacex, const double spacey)
{
    if (_invalid(ppPointFillSymbol)) return false;
    *ppPointFillSymbol = new CPointFillSymbol();
    (*ppPointFillSymbol)->_AddRef();
    (*ppPointFillSymbol)->SetColor(color);
    (*ppPointFillSymbol)->SetPointsSpace(spacex, spacey);
    return true;
}

bool DEF_EXPORT WINAPI CreateSimpleTextSymbol(const double width, const double height,
    const char* const text, const COLORREF textcolor, ISimpleTextSymbol** ppSimpleTextSymbol)
{
    if (_invalid(ppSimpleTextSymbol)) return false;
    assert(!*ppSimpleTextSymbol);

    *ppSimpleTextSymbol = new CSimpleTextSymbol;
    (*ppSimpleTextSymbol)->_AddRef();
    (*ppSimpleTextSymbol)->SetWidth(width);
    (*ppSimpleTextSymbol)->SetHeight(height);
    (*ppSimpleTextSymbol)->SetText(text);
    (*ppSimpleTextSymbol)->SetColor(textcolor);
    return true;
}

bool DEF_EXPORT WINAPI CreateMultiPointSymbol(IMultiPointSymbol** ppMultiPointSymbol)
{
    if (_invalid(ppMultiPointSymbol)) return false;
    assert(!*ppMultiPointSymbol);

    *ppMultiPointSymbol = new CMultiPointSymbol;
    (*ppMultiPointSymbol)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI CreateMultiLineSymbol(IMultiLineSymbol** ppMultiLineSymbol)
{
    if (_invalid(ppMultiLineSymbol)) return false;
    assert(!*ppMultiLineSymbol);

    *ppMultiLineSymbol = new CMultiLineSymbol;
    (*ppMultiLineSymbol)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI CreateMultiFillSymbol(IMultiFillSymbol** ppMultiFillSymbol)
{
    if (_invalid(ppMultiFillSymbol)) return false;
    assert(!*ppMultiFillSymbol);

    *ppMultiFillSymbol = new CMultiFillSymbol;
    (*ppMultiFillSymbol)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI CreateSlimData(const char* const filename, ILayer** ppLayer,
    const MapUnits mapunit, const double basescale, const double precision, const long indexlevel,
    const ShapeType shapetype, const WKSRect extent, const IFields* pFields, const bool anno,
    const bool filemap)
{
    if (_invalid(ppLayer)) return false;
    assert(!*ppLayer);

    IFieldsPtr pIFields = (IFields*)pFields;
    CFieldsPtr pCFields;
    CAST_PTR(pIFields, pCFields, CFields);
    CSlimLayerPtr pSL = new CSlimLayer(shapetype, mapunit, basescale, precision, extent,
        indexlevel, pCFields, anno);
    if (!pSL->Valid())
    {
        return false;
    }

    string sfilename = Trim(filename);
    if (sfilename != "")
    {
        string pathpart = GetDirectoryPart(sfilename);
        CreateDirectory(pathpart.c_str(), NULL);
        pSL->AttachToFile(sfilename, filemap);
    }

    ILayerPtr pLayer;
    CAST_PTR(pSL, pLayer, ILayer)
    *ppLayer = pLayer._p();
    (*ppLayer)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI LoadSlimData(const char* const filename, ILayer** ppLayer, bool readonly,
    bool filemap)
{
    if (_invalid(ppLayer)) return false;
    assert(!*ppLayer);

    CSlimLayerPtr pSL = new CSlimLayer(filename, readonly, filemap);
    if (!pSL->Valid())
    {
        return false;
    }

    ILayerPtr pLayer;
    CAST_PTR(pSL, pLayer, ILayer)
    *ppLayer = pLayer._p();
    (*ppLayer)->_AddRef();
    return true;
}

bool DEF_EXPORT WINAPI LoadShapeFile(const char* const filename, ILayer** ppLayer,
    MapUnits mapunit, double basescale, double precision, long indexlevel, bool readonly)
{
    if (_invalid(ppLayer)) return false;
    assert(!*ppLayer);

    CShapeLayerPtr pSL = new CShapeLayer(filename, mapunit, basescale, precision,
        indexlevel, readonly);
    if (!pSL->Valid())
    {
        return false;
    }

    ILayerPtr pLayer;
    CAST_PTR(pSL, pLayer, ILayer)
    *ppLayer = pLayer._p();
    (*ppLayer)->_AddRef();
    return true;
}

}

File: Delphi SDK/easylib.pas
unit easylib;

interface

uses
    Windows, InterfaceObj, InterfaceGeometry, InterfaceDisplayTransformation, InterfaceDisplay,
    InterfaceSymbol, InterfaceLayer, InterfaceFields, WKSStructs;

type
    //反序列化
    TInstantiate = function(pStream: IStreamX; out pPersist: IPersist): Longword; stdcall;

    //创建对象
    TCreateObj = procedure(classname: PChar; out obj: IObj); stdcall;

    //-------------------------------------------------------------------------------------
    //  为了方便
    //-------------------------------------------------------------------------------------
    TCreateGeometry = function(const geometrytype: TGeometryType;
        out ppGeometry: IGeometry): Boolean; stdcall;
    TCreateSimplePointSymbol = function(const diameter: Double; const color: COLORREF;
        out ppSimplePointSymbol: ISimplePointSymbol): Boolean; stdcall;
    TCreateEnvelopePointSymbol = function(const width: Double; const height: Double; const color: COLORREF;
        out ppEnvelopePointSymbol: IEnvelopePointSymbol): Boolean; stdcall;
    TCreatePolyPointSymbol = function(const solid: Boolean; const linewidth: Double;
        const color: COLORREF; out ppPolyPointSymbol: IPolyPointSymbol): Boolean; stdcall;
    TCreateSimpleLineSymbol = function(const width: Double; const color: COLORREF;
        out ppSimpleLineSymbol: ISimpleLineSymbol): Boolean; stdcall;
    TCreatePointLineSymbol = function(out ppPointLineSymbol: IPointLineSymbol): Boolean; stdcall;
    TCreateSimpleFillSymbol = function(const color: COLORREF; const borderwidth: Double;
        const bordercolor: COLORREF; out ppSimpleFillSymbol: ISimpleFillSymbol): Boolean; stdcall;
    TCreatePointFillSymbol = function(out ppPointFillSymbol: IPointFillSymbol;
        const color: COLORREF; const spacex: Double; const spacey: Double): Boolean; stdcall;
    TCreateSimpleTextSymbol = function(const width: Double; const height: Double; text: PChar;
        const textcolor: COLORREF; out ppSimpleTextSymbol: ISimpleTextSymbol): Boolean; stdcall;
    TCreateMultiPointSymbol = function(out ppMultiPointSymbol: IMultiPointSymbol): Boolean; stdcall;
    TCreateMultiLineSymbol = function(out ppMultiLineSymbol: IMultiLineSymbol): Boolean; stdcall;
    TCreateMultiFillSymbol = function(out ppMultiFillSymbol: IMultiFillSymbol): Boolean; stdcall;

    TCreateSlimData = function(const filename: PChar; out pLayer: ILayer; const mapunit: TMapUnits;
        const basescale: Double; precision: Double; indexlevel: Longint; const shapetype: TShapeType;
        const extent: WKSRect; const pFields: IFields; const anno: Boolean; const filemap: Boolean): Boolean; stdcall;
    TLoadSlimData = function(const filename: PChar; out pLayer: ILayer; readonly: Boolean;
        filemap: Boolean): Boolean; stdcall;
    TLoadShapeFile = function(const filename: PChar; out pLayer: ILayer; const mapunit: TMapUnits;
        const basescale: Double; precision: Double; indexlevel: Longint; readonly: Boolean): Boolean; stdcall;
    //-------------------------------------------------------------------------------------

function InitialEasyLib(const easyfile: PChar): Boolean;
function FreeEasyLib(): Boolean;
function LoadSymbolLib(const filename: PChar; pSymbolLib: ISymbolLib): Boolean; overload;
function LoadSymbolLib(const filename: PChar; pSymbolLib: ISymbolLib;
    const symboltype: TSymbolType): Boolean; overload;

var
    easylibmodule: HMODULE;

    Instantiate: TInstantiate;
    CreateObj: TCreateObj;
    CreateGeometry: TCreateGeometry;
    CreateSimplePointSymbol: TCreateSimplePointSymbol;
    CreateEnvelopePointSymbol: TCreateEnvelopePointSymbol;
    CreatePolyPointSymbol: TCreatePolyPointSymbol;
    CreateSimpleLineSymbol: TCreateSimpleLineSymbol;
    CreatePointLineSymbol: TCreatePointLineSymbol;
    CreateSimpleFillSymbol: TCreateSimpleFillSymbol;
    CreatePointFillSymbol: TCreatePointFillSymbol;
    CreateSimpleTextSymbol: TCreateSimpleTextSymbol;
    CreateMultiPointSymbol: TCreateMultiPointSymbol;
    CreateMultiLineSymbol: TCreateMultiLineSymbol;
    CreateMultiFillSymbol: TCreateMultiFillSymbol;
    CreateSlimData: TCreateSlimData;
    LoadSlimData: TLoadSlimData;
    LoadShapeFile: TLoadShapeFile;

implementation

function InitialEasyLib(const easyfile: PChar): Boolean;
begin
    Result := False;
    easylibmodule := LoadLibrary(easyfile);
    if (0 = easylibmodule) then
    begin
        Exit;
    end;

    Instantiate                 := TInstantiate(GetProcAddress(easylibmodule, 'Instantiate'));
    CreateObj                   := TCreateObj(GetProcAddress(easylibmodule, 'CreateObj'));
    CreateGeometry              := TCreateGeometry(GetProcAddress(easylibmodule, 'CreateGeometry'));
    CreateSimplePointSymbol     := TCreateSimplePointSymbol(GetProcAddress(easylibmodule, 'CreateSimplePointSymbol'));
    CreateEnvelopePointSymbol   := TCreateEnvelopePointSymbol(GetProcAddress(easylibmodule, 'CreateEnvelopePointSymbol'));
    CreatePolyPointSymbol       := TCreatePolyPointSymbol(GetProcAddress(easylibmodule, 'CreatePolyPointSymbol'));
    CreateSimpleLineSymbol      := TCreateSimpleLineSymbol(GetProcAddress(easylibmodule, 'CreateSimpleLineSymbol'));
    CreatePointLineSymbol       := TCreatePointLineSymbol(GetProcAddress(easylibmodule, 'CreatePointLineSymbol'));
    CreateSimpleFillSymbol      := TCreateSimpleFillSymbol(GetProcAddress(easylibmodule, 'CreateSimpleFillSymbol'));
    CreatePointFillSymbol       := TCreatePointFillSymbol(GetProcAddress(easylibmodule, 'CreatePointFillSymbol'));
    CreateSimpleTextSymbol      := TCreateSimpleTextSymbol(GetProcAddress(easylibmodule, 'CreateSimpleTextSymbol'));
    CreateMultiPointSymbol      := TCreateMultiPointSymbol(GetProcAddress(easylibmodule, 'CreateMultiPointSymbol'));
    CreateMultiLineSymbol       := TCreateMultiLineSymbol(GetProcAddress(easylibmodule, 'CreateMultiLineSymbol'));
    CreateMultiFillSymbol       := TCreateMultiFillSymbol(GetProcAddress(easylibmodule, 'CreateMultiFillSymbol'));
    LoadSlimData                := TLoadSlimData(GetProcAddress(easylibmodule, 'LoadSlimData'));
    LoadShapeFile               := TLoadShapeFile(GetProcAddress(easylibmodule, 'LoadShapeFile'));

    Result := True;
end;

function FreeEasyLib(): Boolean;
begin
    Result := False;
    if (0 = easylibmodule) then begin
        Exit;
    end;
    
    FreeLibrary(easylibmodule);
    Result := True;
end;

function LoadSymbolLib(const filename: PChar; pSymbolLib: ISymbolLib): Boolean;
var
    FindFileData: WIN32_FIND_DATA;
    pObj: IObj;
    pSL: ISymbolLib;
    pSymbol: ISymbol;
    i, n: Longint;
    desc: AnsiString;
begin
    Result := False;

    if (not Assigned(pSymbolLib)) then begin
        Exit;
    end;

    if (FindFirstFile(filename, FindFileData) = INVALID_HANDLE_VALUE) then begin
        Exit;
    end;

    CreateObj('CSymbolLib', pObj);
    pSL := ISymbolLib(GotoInterface(pObj, 'ISymbolLib'));
    pSL.LoadFromFile(filename);

    pSymbolLib.ClearSymbols();
    desc := pSL.GetDesc();
    pSymbolLib.SetDesc(PChar(desc));
    n := pSL.GetSymbolCount();
    for i := 0 to n - 1 do begin
        pSL.GetSymbolRef(pSymbol, i);
        pSymbolLib.AddSymbol(pSymbol);
    end;

    Result := True;
end;

function LoadSymbolLib(const filename: PChar; pSymbolLib: ISymbolLib;
    const symboltype: TSymbolType): Boolean;
var
    pObj: IObj;
    pSL: ISymbolLib;
    pSymbol: ISymbol;
    i, n: Longint;
begin
    CreateObj('CSymbolLib', pObj);
    pSL := ISymbolLib(GotoInterface(pObj, 'ISymbolLib'));
    Result := LoadSymbolLib(filename, pSL);
    if (not Result) then Exit;

    pSymbolLib.ClearSymbols();
    n := pSL.GetSymbolCount();
    for i := 0 to n - 1 do begin
        pSL.GetSymbolRef(pSymbol, i);
        if (pSymbol.GetSymbolType() <> symboltype) then begin
            Continue;
        end;

        pSymbolLib.AddSymbol(pSymbol);
    end;

    Result := True;
end;

end.
